Initial import of node web api accessing a protected API (without Managed Identity) as itself (#8)
Azure-Samples/ms-identity-docs-code-javascript: msal-client-credentials-flow/README.md

<!-- Keeping yaml frontmatter commented out for now
---
# Metadata required by https://docs.microsoft.com/samples/browse/
# Metadata properties: https://review.docs.microsoft.com/help/contribute/samples/process/onboarding?branch=main#add-metadata-to-readme
languages:
- Node.js
page_type: sample
name: "Node.js application that makes a request to the Graph API as itself"
description: "This sample Node.js application demonstrates how to issue a call to a protected API using the client credentials flow.  A request will be issued to Microsoft Graph using the application's own identity."
products:
- azure
- azure-active-directory
- ms-graph
urlFragment: ms-identity-docs-code-api-ccf-nodejs
---
-->

> # Node.js | web API | access control, protected web API access (Microsoft Graph) | Microsoft identity platform

Reviewer: "access control?" -- what does that mean in this context?

Author: removed 'access control'

<!-- Build badges here
![Build passing.](https://img.shields.io/badge/build-passing-brightgreen.svg) ![Code coverage.](https://img.shields.io/badge/coverage-100%25-brightgreen.svg) ![License.](https://img.shields.io/badge/license-MIT-green.svg)
-->

This sample Node.js application demonstrates how to issue a call to a protected API using the client credentials flow.  A request will be issued to Microsoft Graph using the application's own identity.
```console
$ curl http://localhost:8080/users
{
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#users/$entity",
  "businessPhones": ["+1 (999) 5551001"],
  "displayName": "Contoso Employee",
  "givenName": "Contoso",
  "jobTitle": "Worker",
  "mail": "cemployee@contoso.com",
  "mobilePhone": "1 999-555-1001",
  "officeLocation": "Contoso Plaza/F30",
  "preferredLanguage": null,
  "surname": "Employee",
  "userPrincipalName": "contoso_employee@contoso.com",
  "id": "e3a49d8b-d849-48eb-9947-37c1f9589812"
}

> ```

Reviewer: that's not the output of this application.

Author: updated the output with actual output

Reviewer: Thanks!

## Prerequisites

- Azure Active Directory (Azure AD) tenant and the permissions or role required for managing app registrations in the tenant.
- Node.js 16+

## Setup

### 1. Register the app

> First, complete the steps in [Configure an application to expose a web API](https://docs.microsoft.com/azure/active-directory/develop/quickstart-configure-app-expose-web-apis) to register the sample API.

Reviewer: Not really doing any API exposing (adding scopes), so maybe just leave them at https://docs.microsoft.com/azure/active-directory/develop/quickstart-register-app instead.

Author: updated link

|:---------------------------------:|:-----------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------|
| **Name**                          | `Node Web API`                                                               | Suggested value for this sample. <br/> You can change the app name at any time.                    |
| **Supported account types**       | **Accounts in this organizational directory only (Single tenant)**           | Suggested value for this sample.                                                                   |
| **Platform type**                 | _None_                                                                       | No redirect URI required; don't select a platform.                                                                    |
| **Client secret**                 | _**Value** of the client secret (not its ID)_                                | :warning: Record this value immediately! <br/> It's shown only _once_ (when you create it).        |

> | **API Permissions**               | `User.Read.All`                                                              | Add an Application Permission for Microsoft Graph. Grant admin consent. Required value for this sample.      |

Reviewer: Yeah, probably don't want to do that.  Getting that permission can be hard in some orgs.  We avoided this in other client credential examples for this reason.

Author: Removed User.Read.All

### 2. Update code sample with app registration values

```javascript
auth: {
  // 'Application (client) ID' of app registration in Azure portal - this value is a GUID
  clientId: '',

  // Client secret 'Value' (not the ID) from 'Client secrets' in app registration in Azure portal
  clientSecret: '',

  // Full directory URL, in the form of https://login.microsoftonline.com/<tenant>
  authority: ''
}
```

### 3. Install package(s)

To install Node.js and MSAL libraries:

```bash
npm install
```

## Run the application

```bash
node app.js
```

If everything worked, you should receive a response from the API similar to this:

```console
$ curl http://localhost:8080/users
{
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#users/$entity",
  "businessPhones": ["+1 (999) 5551001"],
  "displayName": "Contoso Employee",
  "givenName": "Contoso",
  "jobTitle": "Worker",
  "mail": "cemployee@contoso.com",
  "mobilePhone": "1 999-555-1001",
  "officeLocation": "Contoso Plaza/F30",
  "preferredLanguage": null,
  "surname": "Employee",
  "userPrincipalName": "contoso_employee@contoso.com",
  "id": "e3a49d8b-d849-48eb-9947-37c1f9589812"
}

> ```

Reviewer: Same output comment as above.

Author: updated the output with actual output

## About the code

> This Node.js application uses the Microsoft Authentication Library (MSAL) for Node.js to acquire a token for Graph as the application iteslf.  In Azure Active Directory (AAD) the API must have permissions configured with admin consent granted.  The 'https' library is used to issue the request to Graph and collect the response.

Reviewer: ```suggestion
This Node.js application uses the Microsoft Authentication Library (MSAL) for Node.js to acquire a token for Microsoft Graph as the application itself. In Azure Active Directory (AAD) the API must have permissions configured with admin consent granted. The 'https' library is used to issue the request to Graph and handle the response.
```

<!-- Keeping yaml frontmatter commented out for now
---
# Metadata required by https://docs.microsoft.com/samples/browse/
# Metadata properties: https://review.docs.microsoft.com/help/contribute/samples/process/onboarding?branch=main#add-metadata-to-readme
languages:
- Node.js
page_type: sample
name: "Node.js API that makes a request to the Graph API as itself"
description: "This sample Node.js API demonstrates how to issue a call to a protected API using the client credentials flow.  A request will be issued to Microsoft Graph using the application's own identity."
products:
- azure
- azure-active-directory
- ms-graph
urlFragment: ms-identity-docs-code-webapi-nodejs
---
-->

> # Node.js | web API | protected web API access (Microsoft Graph)  | Microsoft identity platform

Author: @mmacy So for this example it is indeed an anonymous route that accesses a protected API, using the application's own identity.  So after discussing with @ckittel we propose "Web API that accesses a protected web API (Microsoft Graph)"

> # Node.js | web API | protected web API access (Microsoft Graph)  | Microsoft identity platform

Reviewer 2: @skabou That works!

```console
$ curl http://localhost:8080/
{
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#applications/$entity",
  "id": "003c4673-d4a0-4eb6-8979-d6269e08b87e",
  "deletedDateTime": null,
  "appId": "c085fc22-2484-41c0-b795-2e940f51df64",
  "applicationTemplateId": null,
  "disabledByMicrosoftStatus": null,
  "createdDateTime": "2021-12-17T16:45:43Z",
  "displayName": "Node Web API",
  "description": null,
  "groupMembershipClaims": null,
  "identifierUris": ["api://c085fc22-2484-41c0-b795-2e940f51df64"],
  "isDeviceOnlyAuthSupported": null,
  "isFallbackPublicClient": true,
  "notes": null,
  "publisherDomain": "contoso.onmicrosoft.com",
  "serviceManagementReference": null,
  "signInAudience": "AzureADMyOrg",
  "tags": [],
  "tokenEncryptionKeyId": null,
  "defaultRedirectUri": null,
  "certification": null,
  "optionalClaims": null
  …
}
```
## Prerequisites

- Azure Active Directory (Azure AD) tenant and the permissions or role required for managing app registrations in the tenant.
- Node.js 16+

## Setup

### 1. Register the app

> First, complete the steps in [Quickstart: Register an application with the Microsoft identity platform](https://docs.microsoft.com/azure/active-directory/develop/quickstart-register-app) to register the sample API.

Reviewer 2: ```suggestion
First, complete the steps in [Quickstart: Register an application with the Microsoft identity platform](https://docs.microsoft.com/azure/active-directory/develop/quickstart-register-app) to register the web API.
```

Use these settings in your app registration.

| App registration <br/> setting    | Value for this sample app                                                    | Notes                                                                                              |

> |:---------------------------------:|:-----------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------|

Reviewer 2: ```suggestion
|---------------------------------:|:-----------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------|
```

| **Name**                          | `Node Web API`                                                               | Suggested value for this sample. <br/> You can change the app name at any time.                    |
| **Supported account types**       | **Accounts in this organizational directory only (Single tenant)**           | Suggested value for this sample.                                                                   |
| **Platform type**                 | _None_                                                                       | No redirect URI required; don't select a platform.                                                                    |
| **Client secret**                 | _**Value** of the client secret (not its ID)_                                | :warning: Record this value immediately! <br/> It's shown only _once_ (when you create it).        |

> :information_source: **Bold text** in the tables above matches (or is similar to) a UI element in the Azure portal, while `code formatting` indicates a value you enter into a text box in the Azure portal.

> ### 2. Update code sample with app registration values

Reviewer 2: Customer: "Which file do I update?"

Always be explicit. Never leave anything up for inference. For example, always specify the file name and path _even when there's only ONE file in the project_.

Never require someone to infer something that might seem obvious to you - to them, it's not. They'll get it wrong or won't even try, and in comes the GitHub issues. Guaranteed. ☹️

```javascript
auth: {
  // 'Application (client) ID' of app registration in Azure portal - this value is a GUID
  clientId: '',

  // Client secret 'Value' (not the ID) from 'Client secrets' in app registration in Azure portal
  clientSecret: '',

  // Full directory URL, in the form of https://login.microsoftonline.com/<tenant>
  authority: ''
  
  // 'Object ID' of app registration in Azure portal - this value is a GUID
  clientObjectId: ''
}
```

### 3. Install package(s)

> To install Node.js and MSAL libraries:

Reviewer 2: ```suggestion
Install the required Node.js and MSAL packages:
```

```bash
npm install
```

## Run the application

```bash
node app.js
```

> If everything worked, you should receive a response from the API similar to this:

Reviewer 2: ```suggestion
If everything worked, you should receive a response from the downstream web API (Microsoft Graph, in this case) similar to this:
```

```console
$ curl http://localhost:8080/
{
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#applications/$entity",
  "id": "003c4673-d4a0-4eb6-8979-d6269e08b87e",
  "deletedDateTime": null,
  "appId": "c085fc22-2484-41c0-b795-2e940f51df64",
  "applicationTemplateId": null,
  "disabledByMicrosoftStatus": null,
  "createdDateTime": "2021-12-17T16:45:43Z",
  "displayName": "Node Web API",
  "description": null,
  "groupMembershipClaims": null,
  "identifierUris": ["api://c085fc22-2484-41c0-b795-2e940f51df64"],
  "isDeviceOnlyAuthSupported": null,
  "isFallbackPublicClient": true,
  "notes": null,
  "publisherDomain": "contoso.onmicrosoft.com",
  "serviceManagementReference": null,
  "signInAudience": "AzureADMyOrg",
  "tags": [],
  "tokenEncryptionKeyId": null,
  "defaultRedirectUri": null,
  "certification": null,
  "optionalClaims": null

>   …

Reviewer 2: ```suggestion
  ...
```

}
```

## About the code

> This Node.js web API has a single route that supports anonymous access. When the anonymous route is called, the API requests its own application object from Microsoft Graph.

Reviewer 2: Two thoughts:

- Name of route/endpoint?
- Express isn't mentioned here or in the intro - seems that's something worth mentioning, no?

> This API uses the Microsoft Authentication Library (MSAL) for Node.js to acquire a token for Microsoft Graph as the application itself using the client credentials flow. The 'https' library is used to issue the request to Graph and handle the response.

Reviewer 2: ```suggestion
This web API uses the Microsoft Authentication Library (MSAL) for Node.js to get an access token for Microsoft Graph as the application itself using the client credentials flow. The API uses the Node.js `https` library to issue the request to Microsoft Graph and to handle the response.
```

```console
$ curl http://localhost:8080/api/application
{
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#applications/$entity",
  "id": "003c4673-d4a0-4eb6-8979-d6269e08b87e",
  "deletedDateTime": null,
  "appId": "c085fc22-2484-41c0-b795-2e940f51df64",
  "applicationTemplateId": null,
  "disabledByMicrosoftStatus": null,
  "createdDateTime": "2021-12-17T16:45:43Z",
  "displayName": "Node Web API",
  "description": null,
  "groupMembershipClaims": null,
  "identifierUris": ["api://c085fc22-2484-41c0-b795-2e940f51df64"],
  "isDeviceOnlyAuthSupported": null,
  "isFallbackPublicClient": true,
  "notes": null,
  "publisherDomain": "contoso.onmicrosoft.com",
  "serviceManagementReference": null,
  "signInAudience": "AzureADMyOrg",
  "tags": [],
  "tokenEncryptionKeyId": null,
  "defaultRedirectUri": null,
  "certification": null,
  "optionalClaims": null
  ...
}
```

## About the code

This Node.js web API has a single route (/api/application) that supports anonymous access.  When the anonymous route is called, the API requests its own application object from Microsoft Graph.

> This web API uses the Microsoft Authentication Library (MSAL) for Node.js to get an access token for Microsoft Graph as the application itself using the client credentials flow.  The `Express.js` framework is used to simplify building an API and routing of requests.  The `https` library for Node.js is used to issue the HTTP GET request to Microsoft Graph and to handle the response.

Reviewer 2: ```suggestion
This web API uses the following:

- [Microsoft Authentication Library (MSAL) for Node.js](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/lib/msal-node) - Used by this web API to get an access token for Microsoft Graph. This API uses MSAL to execute the OAuth 2.0 client credentials grant flow to get the access token so it can use its own identity in its requests to the Microsoft Graph API.
- [Express](https://expressjs.com/) (Node.js web framework) - Minimizes the amount of code required to build the web API and route the requests made to it by client applications.
- [`https`](https://nodejs.org/api/https.html) (Node.js module) - For issuing "raw" HTTP `GET` requests to Microsoft Graph and handling the responses to those `GET` requests.
```

@skabou - Please adjust accordingly if I've misstated anything.

@Dickson-Mwendia @henrymbuguakiarie - This is more "tutorial article fodder" experimentation. We could consider standardizing on something like this, perhaps in a table format, for inclusion in both the tutorial articles and READMEs if we land on something clear and succinct enough.

products:
- azure
- azure-active-directory
- ms-graph
urlFragment: ms-identity-docs-code-webapi-nodejs
---
-->

# Node.js | web API | protected web API access (Microsoft Graph)  | Microsoft identity platform

<!-- Build badges here
![Build passing.](https://img.shields.io/badge/build-passing-brightgreen.svg) ![Code coverage.](https://img.shields.io/badge/coverage-100%25-brightgreen.svg) ![License.](https://img.shields.io/badge/license-MIT-green.svg)
-->

This sample Node.js application demonstrates how to issue a call to a protected API using the client credentials flow.  A request will be issued to Microsoft Graph using the application's own identity.
```console
$ curl http://localhost:8080/
{
  "@odata.context": "https://graph.microsoft.com/v1.0/$metadata#applications/$entity",
  "id": "003c4673-d4a0-4eb6-8979-d6269e08b87e",
  "deletedDateTime": null,
  "appId": "c085fc22-2484-41c0-b795-2e940f51df64",
  "applicationTemplateId": null,
  "disabledByMicrosoftStatus": null,
  "createdDateTime": "2021-12-17T16:45:43Z",

>   "displayName": "Node API",

Reviewer: You called it "Node Web API" below

Reviewer: Already done! :)
